### PR TITLE
Fix blender rest export of individual skinned meshes

### PR DESCRIPTION
**Pull request by daniel-nth**
This PR fixes rest export of individual skinned meshes.
- use the right armature to create the bones array for rest export 5c72366a949e8b8a1f4a8d251c7fd1f7a3c9b107
- re-enable rest export in the exporter settings 6794d284611f429f10da8a98e8faa82e66cf51a0
- allow BlendCharacter to either use ObjectLoader or JSONLoader b88011b4d256f582bedeef4147f72119bba465af
- revert 362cb796d004953274e4986520402de79a68cac5 as it breaks export of individual meshes ac82a4ce52d9b5dd575f6168b75a5dae8daba029

You can find a working marine example with code from this PR here: https://github.com/daniel-nth/three.js-rest-example

@phfatmonkey Could you explain your reasoning for 362cb796d004953274e4986520402de79a68cac5? Surely we can find a solution that works for your use case but doesn't break individual meshes.

As mentioned in https://github.com/mrdoob/three.js/issues/6050#issuecomment-150026481, there should probably be a better solution than 5c72366a949e8b8a1f4a8d251c7fd1f7a3c9b107. Ideally the bones array should be identical wether pose or rest is used.

### CODE REVIEW

**Comment by DelvarWorld** [comment]
cc @s9k @phfatmonkey @tschw @carlosanunes anyone else? Multiple people are making major changes to the Blender exporter - until we have a test suite for the exporter (running Blender headless and testing export process) I think more collaboration is better so we're all on the same page.

**Comment by tschw** [comment]
@daniel-nth 
> Surely we can find a solution that works for your use case but doesn't break individual meshes.

:+1: 

Yes, we should re-test and verify that the recent issues are still fixed.

@DelvarWorld 
Thanks for notifying!

> running Blender headless and testing export process

Sounds like you have experience with this kind of setup. Would it be possible for you to provide a simple framework, maybe?

**Comment by daniel-nth** [comment]
> until we have a test suite for the exporter (running Blender headless and testing export process) 

There are tests in https://github.com/mrdoob/three.js/tree/dev/utils/exporters/blender/tests. But they seem to be outdated. I ran test_geometry_animation.bash and the resulting review page uses THREE.AnimationHandler instead of THREE.AnimationMixer and fails. The headless export still works though. We could use that as a base and create new tests.

**Comment by daniel-nth** [comment]
With the code from #8783 I get these results when running the test. (The GIFs are not perfect loops)

**Current dev branch**
dev-branch  

**This PR**
pr-8781

File: examples/js/BlendCharacter.js [hunk 1]
```diff
@@ -7,7 +7,7 @@ THREE.BlendCharacter = function () {
 	this.weightSchedule = [];
 	this.warpSchedule = [];
 
-	this.load = function ( url, onLoad ) {
+	this.loadObject = function ( url, onLoad ) {
```
Comment (mrdoob) [inline examples/js/BlendCharacter.js]: I think I would keep this as `this.load`.

Comment (daniel-nth) [inline examples/js/BlendCharacter.js]: Okay, done.

**Comment by tschw** [comment]
@DelvarWorld summarized most of the recent exporter issues in https://github.com/mrdoob/three.js/issues/8677#issuecomment-213630937.

I think it would be good to go through these tickets and check whether the models still export correctly. If you can create add them to the test suite, even better.

**Comment by phfatmonkey** [comment]
Thanks for getting to this, @daniel-nth. 

> @phfatmonkey Could you explain your reasoning for 362cb79? Surely we can find a solution that works for your use case but doesn't break individual meshes.

@insominx and I work on this for a few weeks a year, and we were under the impression that the JSONLoader is deprecated and we should be using the Object Loader. When we did that all skinned meshes were being flipped on their sides. The fix I made got them to export as expected for the project we needed.

I'd like to test when I get back in the office tomorrow, but I like where you're going with this.

**Comment by mrdoob** [comment]
Merging this for now. @phfatmonkey let us know if you find any issues while testing.

**Comment by mrdoob** [comment]
Thanks!

**Comment by daniel-nth** [comment]
Thanks!

I'll try to write some more tests. We should have at least on test for rest and pose in geometry and scene export mode. Maybe also with a more complex model than the simple box as errors are much more exaggerated in more complex models.

Some progress can be found here: https://github.com/daniel-nth/three.js/tree/more-tests. Scene export and more complex models are missing.
As the pose export was segfaulting blender in headless mode for me, I changed the way the actions are selected. I also try to set the pose/rest mode of the armature automatically based on the selected export method as this is necessary to use the same blend file for both tests. These two changes definitively need some testing and could be horribly wrong.

**Comment by tschw** [comment]
How to verify the new tests? Compare with how the blend file looks in blender? Would it make sense to just diff the simple ones against an expected result? Or would it just cause trouble and better be done with something node-ish? Or could we create an HTML test results page with expected/actual side-by-side?

It'd be really nice if we would find an easy way to detect regressions to prevent what happened in `r76`. The exporter can be used in many different ways and extending just one may break the others...

**Comment by daniel-nth** [comment]
I like the idea of a side-by-side display. That could still be showing the wrong result for both if something on the JS side breaks. But it would be a lot better than now. Good idea!
Maybe a diff could also be run as an indicator and show a message in the side-by-side comparison if it fails . In itself a diff would maybe show too many false positives as something simple as version bump would show an error. (However that works, is FORMAT_VERSION 3 or 4.4 right now?)

**Comment by phfatmonkey** [comment]
@daniel-nth The example you linked to at the top of this post references a lot of scripts intended for the marine example file. Do you happen to have a generic example I can use to test a generic skinned mesh exported without the scene? 

The scene export doesn't work, as before. The objects are still exported on their sides. 

However, an important question that needs to be asked: @mrdoob, going forward, is the ObjectLoader the preferred method of loading assets or should we still be using the JSONLoader? @insominx and I had been moving forward with the ObjectLoader pipeline, thinking that was the way forward. Do you have any insight on that?

**Comment by mrdoob** [comment]
Yes. `ObjectLoader` is the way forward. In fact, I don't think there is much left that it can't do compared to `JSONLoader` so we could start deprecating it.

**Comment by phfatmonkey** [comment]
Ok, thanks.

Great, thanks, @mrdoob for the clarification. So the exporter for this PR doesn't work with  ```ObjectLoader```, and if that's the way forward I would think that should be the priority.

Legacy JSON files don't load because it's looking for ```Object``` in the JSON. Perhaps including that in the export regardless of whether scene is checked or not would be a good start in ensuring all exported files at least load with ```ObjectLoader```.

Another thing is the skinned meshes being flipped on their sides. This is still an issue when exporting with the exporter from this PR and using the ```ObjectLoader```.

@daniel-nth any thoughts on how to move toward supporting ```ObjectLoader``` while maintaining some legacy support for ```JSONLoader```?

**Comment by daniel-nth** [comment]
@mrdoob Thanks for clearing that up.

@phfatmonkey Sorry about that. :/ I added a very simple example at http://daniel-nth.github.io/three.js-rest-example/examples/export-review.html. It is based on the review page created by the exporter test and will play a random animation from the loaded file. But if JSONLoader goes the way of the dodo, it shouldn't matter. 

If ObjectLoader will be the only option in the future, the exporter should be able to export single meshes in object format. Any idea how to easily do this? I might be able to have a look at it tomorrow.  

Files exported as JSONLoader compatible files with older versions of the exporter should just continue to work, so there wouldn't be much work needed for legacy support, right? Or should the exporter retain the ability ability to export to a deprecated format?

**Comment by mrdoob** [comment]
> Or should the exporter retain the ability to export to a deprecated format?

Nah... Once `ObjectLoader` is well supported, I would suggest removing the other code path.

**Comment by phfatmonkey** [comment]
> If ObjectLoader will be the only option in the future, the exporter should be able to export single meshes in object format. Any idea how to easily do this?

I think if you export as a scene you'll see the JSON format needed for the ```ObjectLoader```. (Maybe we don't need to export as a scene?) It seems the transformation matrix used to get the vert and bone transforms doesn't work for the skinned meshes. 

Here's a test scene using the object loader with skinned meshes. It has the exported JSON and the .blend file. 
[objectloader_skinnedmesh_test.zip](https://github.com/mrdoob/three.js/files/250734/objectloader_skinnedmesh_test.zip)

I've shifted to another project for the time being so my time with ThreeJS will be limited over the next few weeks. Either way let me know if I can assist with anything in the meantime.

**Comment by daniel-nth** [comment]
Thanks. I played around with the exporter a bit. Maybe the solution could be to always use the scene exporter, and when "Scene" in Blender isn't checked, only export the selected object. This would have the added bonus of not having the separate codepaths for geometry and scene export.

**Comment by phfatmonkey** [comment]
Yeah, making the Scene export default is probably a good start. Export Selected would be a good option for either scene or single object export.

**Comment by dsavoy** [comment]
Sorry for the noise here (and my terrible English), but I was fighting to export some models with the latest version of the exporter (r77), and I notice some issues that I hope can help you.

(1) Bones without "Deform" option selected in Blender are not being exported in REST mode. It breaks all the parent attributes of the generated JSON bones. Using POSE mode all the bones are exported.

(2) The bone influences on the vertices (index and weight) are exported in an strange order. If someone sets an small maximum number of influences (say 2 or 3), most of them have weight 0. The logical thing would be to sort the influences by weights before export, and than just output the most meaningful.

**Pull request merged**